File: Extensions/Tessa.Extensions.Default.Client/Workplaces/FilteringExtensions/RefSectionExtension.cs
#nullable enable
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Extensions.Default.Shared.Workplaces;
using Tessa.Platform.Collections;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Views;
using Tessa.UI.Views.Extensions;
using Tessa.UI.Views.Workplaces.Tree;
using Tessa.Views;
using Tessa.Views.Metadata;
using Tessa.Views.Workplaces;

namespace Tessa.Extensions.Default.Client.Workplaces
{
    /// <summary>
    /// Расширение на узел рабочего места, которое задаёт список значений RefSection и параметров представлений,
    /// которые используются для узла при его скрытии или отображении в режиме отбора (по троеточию из ссылочных контролов),
    /// независимо от представлений в узле.
    /// </summary>
    /// <remarks>
    /// У расширения есть конфигуратор <see cref="RefSectionExtensionConfigurator"/>.
    /// </remarks>
    public sealed class RefSectionExtension :
        ViewModel<EmptyModel>,
        ITreeItemExtension,
        IWorkplaceExtensionSettingsRestore,
        IWorkplaceFilteringRule
    {
        #region Fields

        private TreeItemFilteringSettings settings = new();

        #endregion

        #region Private Methods

        private static bool ParametersEquals(IEnumerable<RequestParameter>? contextParameters, List<string> settingsParameters) =>
            contextParameters.AsReadOnlyCollection() is not { Count: > 0 } parameters
            || parameters.All(x => ParserNames.Contains(x.Name, settingsParameters));

        private static bool ConditionEquals(IWorkplaceFilteringContext context, ITreeItemFilteringSettings settings) =>
            ParserNames.HasIntersections(settings.RefSections, context.RefSection)
            && ParametersEquals(context.Parameters, settings.Parameters);

        #endregion

        #region ITreeItemExtension Members

        /// <inheritdoc />
        public void Clone(ITreeItem source, ITreeItem cloned, ICloneableContext context)
        {
        }

        /// <inheritdoc />
        public void Initialize(ITreeItem model)
        {
        }

        /// <inheritdoc />
        public void Initialized(ITreeItem model)
        {
        }

        #endregion

        #region IWorkplaceFilteringRule Members

        /// <inheritdoc />
        public ValueTask<CheckingResult> EvaluateAsync(IWorkplaceComponentMetadata metadata, IWorkplaceFilteringContext context) =>
            new(ConditionEquals(context, this.settings) ? CheckingResult.Positive : CheckingResult.Negative);

        #endregion

        #region IWorkplaceExtensionSettingsRestore Members

        /// <inheritdoc />
        public void Restore(Dictionary<string, object?> metadata) =>
            this.settings = metadata.FromSerializedDictionary<TreeItemFilteringSettings>();

        #endregion
    }
}

File: WebClient SDK/src/default/workplaces/refSectionExtension.ts
import { TreeItemExtension } from 'tessa/ui/views/extensions';
import {
  IWorkplaceFilteringRule,
  CheckingResult,
  WorkplaceMetadataComponentSealed,
  WorkplaceFilteringContext
} from 'tessa/views/workplaces';
import { IStorage } from 'tessa/platform/storage';
import { StringHelper } from 'tessa/platform';
import { RequestParameter } from 'tessa/views/metadata';
import { extension } from '@tessa/application';

@extension()
export class RefSectionExtension extends TreeItemExtension implements IWorkplaceFilteringRule {
  public getExtensionName(): string {
    return 'Tessa.Extensions.Default.Client.Workplaces.RefSectionExtension';
  }

  async evaluate(
    _metadata: WorkplaceMetadataComponentSealed,
    context: WorkplaceFilteringContext
  ): Promise<CheckingResult> {
    const settings = new TreeItemFilteringSettings(this.settingsStorage);

    const conditionEquals =
      this.settingsStorage &&
      settings.refSections.some(x => context.refSection?.some(y => StringHelper.equals(x, y))) &&
      RefSectionExtension.parametersEquals(context.parameters, settings.parameters);
    return conditionEquals ? CheckingResult.Positive : CheckingResult.Negative;
  }

  private static parametersEquals(
    contextParameters: RequestParameter[],
    settingsParameters: string[]
  ): boolean {
    if (contextParameters == null || contextParameters.length === 0) {
      return true;
    }

    return contextParameters
      .map(p => p.metadata.alias)
      .every(x => settingsParameters.some(y => StringHelper.equals(x, y)));
  }
}

class TreeItemFilteringSettings {
  constructor(storage: IStorage) {
    this.refSections = storage['RefSections'] || [];
    this.parameters = storage['Parameters'] || [];
  }

  public readonly refSections: string[];

  public readonly parameters: string[];
}
